shared/naturalcrit/codeEditor/helpers/widgets.js

```diff
@@ -0,0 +1,93 @@
+const React = require('react');
+const ReactDOM = require('react-dom');
+const { PATTERNS, FIELD_TYPE } = require('./widget-elements/constants');
+require('./widget-elements/hints/hints.jsx');
+
+module.exports = function(CodeMirror, widgets, cm, setHints) {
+	const hintsEl = document.createElement('ul');
+	hintsEl.id = 'hints';
+	hintsEl.role = 'listbox';
+	hintsEl.ariaExpanded = 'true';
+	hintsEl.className = 'CodeMirror-hints default';
+	hintsEl.style = 'display: none;';
+	document.body.append(hintsEl);
+
+	const { cClass, field } = require('./widget-elements')(CodeMirror, setHints);
+	const widgetOptions = widgets.map((widget)=>({
+		name         : widget.name,
+		pattern      : PATTERNS.widget[widget.type](widget.name),
+		createWidget : (n, node)=>{
+			const parent = document.createElement('div');
+			const classes = (widget.classes || []).map((c, i)=>cClass(cm, n, `{{${widget.name}`, c));
+			const fieldNames = (widget.fields || []).map((f)=>f.name);
+			const fields = (widget.fields || []).map((f, i)=>field(cm, n, f)).filter((f)=>!!f);
+			const { text } = cm.lineInfo(n);
+			const styles = [...text.matchAll(PATTERNS.collectStyles)].map(([_, style])=>{
+				if(fieldNames.includes(style)) return false;
+				return field(cm, n, {
+					name      : style,
+					type      : FIELD_TYPE.STYLE,
+					increment : 5
+				});
+			}).filter((s)=>!!s);
+
+			ReactDOM.render(<React.Fragment>
+				{classes}
+				{fields}
+				{styles}
+			</React.Fragment>, node || parent);
+
+			return node || parent;
+		}
+	}));
+
+	const updateLineWidgets = (n, remove)=>{
+		const { text, widgets } =  cm.lineInfo(n);
+		const widgetOption = widgetOptions.find((option)=>!!text.match(option.pattern));
+		if(!widgetOption) return;
+		if(!!widgets) {
+			for (const widget of widgets) {
+				widgetOption.createWidget(n, widget.node);
+			}
```

Comment:
I may need help understanding exactly what happens here. `widget` is a CodeMirror object, right? And `widget.node` is the HTML element representing that widget? So `createWidget` is just replacing the HTML on the existing widget, right? Perhaps we can rename `createWidget` to something like `updateWidgetHTML` or `updateWidgetElement`?

Reply:
Fair, I'll update it to `render`